In [7]:
import pandas as pd

# read in datasets 
mcsilver_data = pd.read_excel("./in/McSilver_NYCProgramExport_10032024.xlsx")
omh_data = pd.read_excel("./in/OMH_ProgramDirectory_11082024.xlsx")
oasas_treatment_data = pd.read_csv("./in/OASAS_TreatmentPrograms_11082024.csv")
oasas_prevention_data = pd.read_csv("./in/OASAS_PreventionPrograms_11082024.csv")

# 1. Viewing and exploring data

In [8]:
mcsilver_data.head()

ACG id  internal id               type  agency acg id  \
0       6      1700005  Outpatient Clinic            4.0   
1       7      1700006  Outpatient Clinic            4.0   
2       8      1700007  Outpatient Clinic            4.0   
3       9      1700008  Outpatient Clinic            5.0   
4      12      1700011  Outpatient Clinic            6.0   

                              agency  \
0     Counseling Service of E.D.N.Y.   
1     Counseling Service of E.D.N.Y.   
2     Counseling Service of E.D.N.Y.   
3            Dynamic Youth Community   
4  Elmcor Youth and Adult Activities   

                                              name      city state    zip  \
0          Outpatient Service (822)- Remsen Street  Brooklyn    NY  11201   
1    Outpatient Services (822-4)- Joralemon Street  Brooklyn    NY  11201   
2      Outpatient Services (822-4)- Jamaica Avenue   Jamaica    NY  11432   
3    Outpatient Service (822)- Coney Island Avenue  Brooklyn    NY  11230   
4  Outpatient Services (822-4)- Northern Boulevard    Corona    NY  11368   

   county  ... category         sub category  \
0   Kings  ...    Oasas  Outpatient Services   
1   Kings  ...    Oasas  Outpatient Services   
2  Queens  ...    Oasas  Outpatient Services   
3   Kings  ...    Oasas  Outpatient Services   
4  Queens  ...    Oasas  Outpatient Services   

                          populations served program project type nyc ros  \
0  Under 18- OASAS,18-21- OASAS,12-21- OASAS                MCTAC     NYC   
1                               18-21- OASAS                MCTAC     NYC   
2  Under 18- OASAS,18-21- OASAS,12-21- OASAS                MCTAC     NYC   
3                               18-21- OASAS                MCTAC     NYC   
4  Under 18- OASAS,18-21- OASAS,12-21- OASAS                MCTAC     NYC   

  license designation date created  closed closed date draft  
0               OASAS   2017-08-21     NaN         NaN   NaN  
1               OASAS   2017-08-21     NaN         NaN   NaN  
2               OASAS   2017-08-21     NaN         NaN   NaN  
3               OASAS   2017-08-21     NaN         NaN   NaN  
4               OASAS   2017-08-21     NaN         NaN   NaN  

[5 rows x 41 columns]

In [9]:
# we want to merge on omhProgramCode, omhFacilityCode, and OMH/OASAS Agency Code
mcsilver_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2156 entries, 0 to 2155
Data columns (total 41 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   ACG id                               2156 non-null   int64         
 1   internal id                          2156 non-null   int64         
 2   type                                 2154 non-null   object        
 3   agency acg id                        2154 non-null   float64       
 4   agency                               2154 non-null   object        
 5   name                                 2156 non-null   object        
 6   city                                 1510 non-null   object        
 7   state                                2156 non-null   object        
 8   zip                                  1510 non-null   object        
 9   county                               1505 non-null   object        
 10  region omh  

In [10]:
omh_data.head()

Sponsor Name  \
0  1003 Walnut Street/dba Sol Stone Center   
1         163rd Street Improvement Council   
2                               ACMH, Inc.   
3                               ACMH, Inc.   
4                               ACMH, Inc.   

                               Agency Name              Program Name  \
0  1003 Walnut Street/dba Sol Stone Center          Sol Stone Center   
1         163rd Street Improvement Council  Flossie Wilson Residence   
2                               ACMH, Inc.             74th Avenue-A   
3                               ACMH, Inc.              ACMH - HH CM   
4                               ACMH, Inc.       ACMH - HH NonMed CM   

  Population Served    Agency Phone   Program Phone  \
0            Adults  (877) 765-7866  (877) 765-7866   
1            Adults  (347) 978-1139  (347) 978-1139   
2            Adults  (212) 274-8558  (212) 274-8558   
3            Adults  (212) 274-8558  (212) 274-8558   
4            Adults  (212) 274-8558  (212) 274-8558   

              Program Address 1 Program Address 2 Program City Program State  \
0            1003 Walnut Street               NaN       Elmira            NY   
1  Agency Phone: (347) 978-1139               NaN          NaN            NY   
2  Agency Phone: (212) 274-8558               NaN          NaN            NY   
3             545 Eighth Avenue          Room 910     New York            NY   
4             545 Eighth Avenue          Room 610     New York            NY   

   ... Operating Certificate Duration Program County    Program Region  \
0  ...                          36/36        Chemung  Western New York   
1  ...                          36/36          Bronx     New York City   
2  ...                          36/36         Queens     New York City   
3  ...                            NaN       New York     New York City   
4  ...                            NaN       New York     New York City   

                               Program Type Program Category  \
0                   Partial Hospitalization       Outpatient   
1                   SRO Community Residence      Residential   
2                   SRO Community Residence      Residential   
3               Health Home Care Management          Support   
4  Health Home Non-Medicaid Care Management          Support   

       Program Subcategory Sponsor Code Agency Code Facility Code  \
0  Partial Hospitalization       810099       44810          7810   
1        Treatment Program       724099       30510          9233   
2        Treatment Program        27099       18780          7062   
3        Care Coordination        27099       18780          7062   
4        Care Coordination        27099       18780          7062   

   Program Code  
0             1  
1             1  
2             6  
3            10  
4            11  

[5 rows x 23 columns]

In [11]:
omh_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6322 entries, 0 to 6321
Data columns (total 23 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Sponsor Name                     6322 non-null   object 
 1   Agency Name                      6322 non-null   object 
 2   Program Name                     6322 non-null   object 
 3   Population Served                5757 non-null   object 
 4   Agency Phone                     6322 non-null   object 
 5   Program Phone                    6322 non-null   object 
 6   Program Address 1                6322 non-null   object 
 7   Program Address 2                1083 non-null   object 
 8   Program City                     4619 non-null   object 
 9   Program State                    6322 non-null   object 
 10  Program Zip                      4619 non-null   object 
 11  Operating Certificate Required?  6322 non-null   object 
 12  Program Tier        

In [12]:
oasas_treatment_data.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 311 entries, 0 to 310
Data columns (total 34 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   PROVIDER_NAME                     311 non-null    object 
 1   PROVIDER_NUMBER                   311 non-null    int64  
 2   PROVIDER_STREET                   311 non-null    object 
 3   PROVIDER_CITY                     311 non-null    object 
 4   PROVIDER_COUNTY                   311 non-null    object 
 5   PROVIDER_STATE                    311 non-null    object 
 6   PROVIDER_ZIP_CODE                 311 non-null    int64  
 7   PROVIDER_ZIP_CODE_PLUS_4          165 non-null    float64
 8   PROGRAM_NAME                      311 non-null    object 
 9   PROGRAM_NUMBER                    311 non-null    int64  
 10  PROGRAM_COUNTY                    311 non-null    object 
 11  PROGRAM_STREET_ADDRESS            311 non-null    object 
 12  PROGRAM_

In [13]:
oasas_prevention_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63 entries, 0 to 62
Data columns (total 22 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   PROVIDER_NAME                     63 non-null     object 
 1   PROVIDER_NUMBER                   63 non-null     int64  
 2   PROVIDER_STREET                   63 non-null     object 
 3   PROVIDER_CITY                     63 non-null     object 
 4   PROVIDER_COUNTY                   63 non-null     object 
 5   PROVIDER_STATE                    63 non-null     object 
 6   PROVIDER_ZIP_CODE                 63 non-null     int64  
 7   PROVIDER_ZIP_CODE_PLUS_4          11 non-null     float64
 8   PROGRAM_NAME                      63 non-null     object 
 9   PROGRAM_NUMBER                    63 non-null     int64  
 10  PROGRAM_COUNTY                    63 non-null     object 
 11  PROGRAM_STREET_ADDRESS            63 non-null     object 
 12  PROGRAM_CI

# 2. Cleaning data

In [14]:
# rename columns 
mcsilver_data         = mcsilver_data.rename(columns=lambda x:x.lower().replace(" ","_"))
omh_data              = omh_data.rename(columns=lambda x:x.lower().replace(" ","_"))
oasas_treatment_data  = oasas_treatment_data.rename(columns=lambda x:x.lower().replace(" ","_"))
oasas_prevention_data = oasas_prevention_data.rename(columns=lambda x:x.lower().replace(" ","_"))

# strip all column names 
mcsilver_data.columns         = mcsilver_data.columns.str.strip() 
omh_data.columns              = omh_data.columns.str.strip() 
oasas_treatment_data.columns  = oasas_treatment_data.columns.str.strip() 
oasas_prevention_data.columns = oasas_prevention_data.columns.str.strip() 

# 3. Merging McSilver and OMH data

There are 1123/2156 rows in the McSilver dataset that have complete OMH program identification. 
We were able to match 877 addresses. 

In [15]:
# subset columns to the ones we care about 
omh_subset = omh_data[["program_address_1", "program_address_2", 
                       "program_city", "program_state", "program_zip",
                       "agency_code", "facility_code", "program_code",  # omhProgramCode, omhFacilityCode, and OMH/OASAS Agency Code
                       "program_type"]]
omh_subset.head()

program_address_1 program_address_2 program_city program_state  \
0            1003 Walnut Street               NaN       Elmira            NY   
1  Agency Phone: (347) 978-1139               NaN          NaN            NY   
2  Agency Phone: (212) 274-8558               NaN          NaN            NY   
3             545 Eighth Avenue          Room 910     New York            NY   
4             545 Eighth Avenue          Room 610     New York            NY   

  program_zip  agency_code  facility_code  program_code  \
0  14901-1007        44810           7810             1   
1         NaN        30510           9233             1   
2         NaN        18780           7062             6   
3       10018        18780           7062            10   
4       10018        18780           7062            11   

                               program_type  
0                   Partial Hospitalization  
1                   SRO Community Residence  
2                   SRO Community Residence  
3               Health Home Care Management  
4  Health Home Non-Medicaid Care Management

In [16]:
# drop all rows in the McSilver dataset that have any missing keys 
mcsilver_omh_subset = mcsilver_data.dropna(subset=["omhprogramcode", "omhfacilitycode", "omh/oasas_agency_code"], how="any")
print(mcsilver_omh_subset.shape)

(1123, 41)


In [17]:
# merge datasets 
mcsilver_omh_merged = mcsilver_omh_subset.merge(omh_subset, 
                                                left_on=["omhprogramcode", "omhfacilitycode", "omh/oasas_agency_code"], 
                                                right_on=["program_code", "facility_code", "agency_code"], 
                                                suffixes=(False, False),
                                                how="left")
print(mcsilver_omh_merged.shape) 
mcsilver_omh_merged["program_address_1"].isna().sum() # 246 missing addresses, 877 matches

# drop extra id variables 
mcsilver_omh_merged.drop(["program_code", "facility_code", "agency_code"], axis=1, inplace=True)

(1123, 50)


# 4. Merging McSilver and OASAS data

There are 321/2156 rows in the McSilver dataset that have complete OASAS program identification. 
We were able to match 234 addresses. 

In [18]:
# stack two OASAS datasets 
oasas_data = pd.concat([oasas_treatment_data, oasas_prevention_data], axis=0)
print(oasas_data.shape)

(374, 34)


In [19]:
# subset columns to the ones we care about 
oasas_subset = oasas_data[["provider_number", "program_number", "service_type",
                           "program_street_address", "program_city", "program_county", "program_state", "program_zip_code"]]

# rename columns to match OMH data 
oasas_subset = oasas_subset.rename(columns={"program_street_address": "program_address_1", 
                                            "program_zip_code": "program_zip",
                                            "service_type": "program_type"})
oasas_subset.head()

provider_number  program_number                    program_type  \
0            10160             227               Outpatient Clinic   
1            10160           53152     820 Rehab and Reintegration   
2            10160           53153     820 Rehab and Reintegration   
3            32412           52100               Outpatient Clinic   
4             1857           53297  820 Stabil Rehab Reintegration   

   program_address_1 program_city program_county program_state  program_zip  
0     830 Forest Ave        Bronx          Bronx            NY        10456  
1     830 Forest Ave        Bronx          Bronx            NY        10456  
2     830 Forest Ave        Bronx          Bronx            NY        10456  
3    3584 Jerome Ave        Bronx          Bronx            NY        10467  
4  1064 Franklin Ave        Bronx          Bronx            NY        10456

In [20]:
# drop all rows in the McSilver dataset that have any missing keys 
mcsilver_oasas_subset = mcsilver_data.dropna(subset=["oasasprogramcode", "omh/oasas_agency_code"], how="any")
print(mcsilver_oasas_subset.shape)

(321, 41)


In [21]:
# merge datasets 
mcsilver_oasas_merged = mcsilver_oasas_subset.merge(oasas_subset, 
                                                   left_on=["oasasprogramcode", "omh/oasas_agency_code"], 
                                                   right_on=["program_number", "provider_number"], 
                                                   suffixes=(False, False),
                                                   how="left")
print(mcsilver_oasas_merged.shape)  
mcsilver_oasas_merged["program_address_1"].isna().sum() # 93 missing addresses, 234 matches

# drop extra id variables 
mcsilver_oasas_merged.drop(["program_number", "provider_number"], axis=1, inplace=True)

(327, 49)


# 5. Other 

After doing the initial merge of OMH and OASAS data, we have 877+234 = 1111/2156 address matches. 
The complete OMH and OASAS data account for 1123+327 = 1450/2156 of the programs in the McSilver data. 

In [22]:
# stack dataframes 
merged_data = pd.concat([mcsilver_omh_merged, mcsilver_oasas_merged], ignore_index=True, axis=0) 

# add flag for whether we were able to find the address
merged_data["matched_address"] = merged_data["program_address_1"].notna()
merged_data["matched_address"].sum()

# write file to Excel 
merged_data.to_excel("./out/mcsilver_omh_oasas_address_data.xlsx")

In [23]:
# get all other rows that we haven't looked at 
remaining_data = pd.merge(mcsilver_data, merged_data["internal_id"], how="left",indicator=True, on="internal_id")
remaining_data = remaining_data[(remaining_data._merge == "left_only")]

remaining_data.head()

# write to file to inspect 
remaining_data.to_excel("./out/remaining_data.xlsx")